File: web/src/api/groups.ts
export interface GroupSummary {
  jid: string;
  name?: string;
  participant_count?: number;
  announce?: boolean;
  locked?: boolean;
  created_at?: string;
}

export interface GroupParticipant {
  jid: string;
  is_admin?: boolean;
  is_super?: boolean;
}

export interface GroupDetail {
  jid: string;
  name?: string;
  topic?: string;
  participant_count?: number;
  announce?: boolean;
  locked?: boolean;
  created_at?: string;
  participants?: GroupParticipant[];
}

async function parseJson<T>(res: Response): Promise<T> {
  const body = (await res.json().catch(() => ({}))) as T & { detail?: string };
  if (!res.ok) {
    throw new Error(
      typeof body.detail === "string" ? body.detail : `Request failed (${res.status})`,
    );
  }
  return body;
}

export async function fetchGroups(): Promise<GroupSummary[]> {
  const res = await fetch("/api/whatsapp/groups", { credentials: "include" });
  const data = await parseJson<{ groups?: GroupSummary[] }>(res);
  return data.groups ?? [];
}

export async function fetchGroup(jid: string): Promise<GroupDetail> {
  const res = await fetch(`/api/whatsapp/groups/${encodeURIComponent(jid)}`, {
    credentials: "include",
  });
  const data = await parseJson<{ group?: GroupDetail; ok?: boolean }>(res);
  return data.group ?? (data as unknown as GroupDetail);
}

export async function createGroup(
  name: string,
  participants: string[],
): Promise<Record<string, unknown>> {
  const res = await fetch("/api/whatsapp/groups", {
    method: "POST",
    credentials: "include",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ name, participants }),
  });
  return parseJson(res);
}

export async function updateGroup(
  jid: string,
  patch: { name?: string; topic?: string; announce?: boolean; locked?: boolean },
): Promise<Record<string, unknown>> {
  const res = await fetch(`/api/whatsapp/groups/${encodeURIComponent(jid)}`, {
    method: "PATCH",
    credentials: "include",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(patch),
  });
  return parseJson(res);
}

export async function updateGroupParticipants(
  jid: string,
  participants: string[],
  action: "add" | "remove" | "promote" | "demote",
): Promise<Record<string, unknown>> {
  const res = await fetch(
    `/api/whatsapp/groups/${encodeURIComponent(jid)}/participants`,
    {
      method: "POST",
      credentials: "include",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ participants, action }),
    },
  );
  return parseJson(res);
}

export async function fetchGroupInvite(
  jid: string,
  reset = false,
): Promise<{ invite_link?: string }> {
  const res = await fetch(`/api/whatsapp/groups/${encodeURIComponent(jid)}/invite`, {
    method: "POST",
    credentials: "include",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ reset }),
  });
  return parseJson(res);
}

export async function joinGroup(inviteLink: string): Promise<Record<string, unknown>> {
  const res = await fetch("/api/whatsapp/groups/join", {
    method: "POST",
    credentials: "include",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ invite_link: inviteLink }),
  });
  return parseJson(res);
}

export async function leaveGroup(jid: string): Promise<Record<string, unknown>> {
  const res = await fetch(`/api/whatsapp/groups/${encodeURIComponent(jid)}/leave`, {
    method: "POST",
    credentials: "include",
  });
  return parseJson(res);
}

export async function setGroupPicture(
  jid: string,
  file: File,
): Promise<Record<string, unknown>> {
  const form = new FormData();
  form.append("file", file);
  const res = await fetch(`/api/whatsapp/groups/${encodeURIComponent(jid)}/picture`, {
    method: "POST",
    credentials: "include",
    body: form,
  });
  return parseJson(res);
}

export async function removeGroupPicture(jid: string): Promise<Record<string, unknown>> {
  const res = await fetch(`/api/whatsapp/groups/${encodeURIComponent(jid)}/picture`, {
    method: "DELETE",
    credentials: "include",
  });
  return parseJson(res);
}

File: src/wabot_agent/api/routes/groups.py
"""WhatsApp group management routes.

Carved out of api/__init__.py as part of MASTER ME-1 Part 4.
"""

from __future__ import annotations

import secrets
from pathlib import Path
from typing import Annotated, Any

from fastapi import APIRouter, Depends, File, HTTPException, UploadFile

from ...auth import verify_human_factory
from ..dependencies import _wabot_call
from ..deps import AppDeps
from ..schemas import (
    GroupCreateRequest,
    GroupInviteRequest,
    GroupJoinRequest,
    GroupParticipantsRequest,
    GroupUpdateRequest,
)


def register_groups_routes(router: APIRouter, deps: AppDeps) -> None:
    settings = deps.settings
    wabot = deps.wabot

    verify_human = verify_human_factory(settings)
    human_dependency = Depends(verify_human)

    @router.get("/api/whatsapp/groups", dependencies=[human_dependency])
    async def list_whatsapp_groups_api() -> dict[str, Any]:
        return await _wabot_call(wabot.list_groups())

    @router.post("/api/whatsapp/groups", dependencies=[human_dependency])
    async def create_whatsapp_group_api(body: GroupCreateRequest) -> dict[str, Any]:
        return await _wabot_call(wabot.create_group(body.name, body.participants))

    @router.get("/api/whatsapp/groups/{group_jid}", dependencies=[human_dependency])
    async def get_whatsapp_group_api(group_jid: str) -> dict[str, Any]:
        return await _wabot_call(wabot.get_group(group_jid))

    @router.patch("/api/whatsapp/groups/{group_jid}", dependencies=[human_dependency])
    async def update_whatsapp_group_api(
        group_jid: str, body: GroupUpdateRequest
    ) -> dict[str, Any]:
        if (
            body.name is None
            and body.topic is None
            and body.announce is None
            and body.locked is None
        ):
            raise HTTPException(
                status_code=400,
                detail="provide at least one of name, topic, announce, locked",
            )
        return await _wabot_call(
            wabot.update_group(
                group_jid,
                name=body.name,
                topic=body.topic,
                announce=body.announce,
                locked=body.locked,
            )
        )

    @router.post(
        "/api/whatsapp/groups/{group_jid}/participants",
        dependencies=[human_dependency],
    )
    async def update_whatsapp_group_participants_api(
        group_jid: str, body: GroupParticipantsRequest
    ) -> dict[str, Any]:
        return await _wabot_call(
            wabot.update_group_participants(
                group_jid, body.participants, action=body.action
            )
        )

    @router.post("/api/whatsapp/groups/{group_jid}/leave", dependencies=[human_dependency])
    async def leave_whatsapp_group_api(group_jid: str) -> dict[str, Any]:
        return await _wabot_call(wabot.leave_group(group_jid))

    @router.post(
        "/api/whatsapp/groups/{group_jid}/picture",
        dependencies=[human_dependency],
    )
    async def set_whatsapp_group_picture_api(
        group_jid: str,
        file: Annotated[UploadFile, File()],
    ) -> dict[str, Any]:
        suffix = Path(file.filename or "group.jpg").suffix or ".jpg"
        tmp = settings.media_dir / f"group-picture-upload-{secrets.token_hex(8)}{suffix}"
        try:
            data = await file.read()
            if not data:
                raise HTTPException(status_code=400, detail="empty file")
            tmp.write_bytes(data)
            return await _wabot_call(wabot.set_group_picture(group_jid, str(tmp)))
        finally:
            tmp.unlink(missing_ok=True)

    @router.delete(
        "/api/whatsapp/groups/{group_jid}/picture",
        dependencies=[human_dependency],
    )
    async def remove_whatsapp_group_picture_api(group_jid: str) -> dict[str, Any]:
        return await _wabot_call(wabot.remove_group_picture(group_jid))

    @router.post(
        "/api/whatsapp/groups/{group_jid}/invite",
        dependencies=[human_dependency],
    )
    async def get_whatsapp_group_invite_api(
        group_jid: str, body: GroupInviteRequest
    ) -> dict[str, Any]:
        return await _wabot_call(wabot.get_group_invite(group_jid, reset=body.reset))

    @router.post("/api/whatsapp/groups/join", dependencies=[human_dependency])
    async def join_whatsapp_group_api(body: GroupJoinRequest) -> dict[str, Any]:
        return await _wabot_call(wabot.join_group(body.invite_link))
